INT-4397: Fix headers filtering for @Transformer

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4397

The `AbstractMessageProcessingTransformer` doesn't honor a configured
`notPropagatedHeaders` and copies all the request headers to the
message to return

* Add `setNotPropagatedHeaders()` into the `AbstractMessageProcessingTransformer`
and implement there a logic to filter headers, similar to what we have
in the `AbstractMessageProducingHandler`
* Overrider `updateNotPropagatedHeaders()` in the `MessageTransformingHandler`
to propagate `notPropagatedHeaders` to the `AbstractMessageProcessingTransformer`
delegate

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
INT-4397: Fix headers filtering for @Transformer

JIRA: https://jira.spring.io/browse/INT-4397

The `AbstractMessageProcessingTransformer` doesn't honor a configured
`notPropagatedHeaders` and copies all the request headers to the
message to return

* Add `setNotPropagatedHeaders()` into the `AbstractMessageProcessingTransformer`
and implement there a logic to filter headers, similar to what we have
in the `AbstractMessageProducingHandler`
* Overrider `updateNotPropagatedHeaders()` in the `MessageTransformingHandler`
to propagate `notPropagatedHeaders` to the `AbstractMessageProcessingTransformer`
delegate

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -139,7 +139,7 @@ public void setNotPropagatedHeaders(String... headers) {
 	 * @since 5.0.2
 	 */
-	protected final void updateNotPropagatedHeaders(String[] headers, boolean merge) {
+	protected void updateNotPropagatedHeaders(String[] headers, boolean merge) {
```
Well, I see now JavaDocs. Not sure why I have missed to read them :cry: 

The problem that this method nowhere is used from the ctor...

What am I missing? Maybe we should just relax this method and let end-user to deal with it on its own?

I think otherwise I will override the setter and adder in the `MessageTransformingHandler`.

Thanks

artembilan replied:
Wait, but those headers are not mapped by the `AbstractKafkaHeaderMapper` anyway.
Don't we use there a `headerMapper` for gateway reply messages?

garyrussell replied:
It has nothing to do with header mapping either outbound or reply; those headers might be present on the outbound message received by the gateway, and we don't want them to appear in the reply message. Hence they are added to the `notPropagatedHeaders`.

artembilan committed:
* Revert `final` for the `AbstractMessageProducingHandler.updateNotPropagatedHeaders()`
* Override `addNotPropagatedHeaders()` for the `MessageTransformingHandler()`
and populate `notPropagatedHeaders` into the target `AbstractMessageProcessingTransformer`
from there
* Also populate `notPropagatedHeaders` from the `AbstractMessageProcessingTransformer.doInit()`
* Implement a `AbstractIntegrationMessageBuilder.filterAndCopyHeadersIfAbsent()`
for a general logic to filter `notPropagatedHeaders` and copy the result
headers set into the target message if they are absent
* Use an new `filterAndCopyHeadersIfAbsent()` in the `AbstractMessageProducingHandler`
and `AbstractMessageProcessingTransformer`to avoid code block duplication
